### PR TITLE
Dto factory form data

### LINKED ISSUE #1529: DTO Factory: form data handling
Need to have PR #1511 signed off.

- [ ] tests
- [ ] review

There is a To Do for documenting this in #1520

### PR DESCRIPTION
**Pull request by peterschutt**
Closes #1529 

### Pull Request Checklist

[//]: # "Please review the [Litestar contribution guidelines](https://github.com/litestar-org/litestar/blob/main/CONTRIBUTING.rst) for this repository."

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

By submitting this issue, you agree to:

- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)
- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)

### Description

[//]: # "Please describe your pull request for new release changelog purposes"

### Close Issue(s)

[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

### CODE REVIEW

File: litestar/_kwargs/kwargs_model.py [hunk 1]
```diff
@@ -438,8 +438,8 @@ def _validate_dependency_data(
                 "Dependencies have incompatible 'data' kwarg types: one expects JSON and the other expects form-data"
             )
         if expected_form_data and dependency_kwargs_model.expected_form_data:
-            local_media_type, _ = expected_form_data
-            dependency_media_type, _ = dependency_kwargs_model.expected_form_data
+            local_media_type, _, _ = expected_form_data
+            dependency_media_type, _, _ = dependency_kwargs_model.expected_form_data
```
Comment (provinzkraut) [inline litestar/_kwargs/kwargs_model.py:442]:
```suggestion
            local_media_type = expected_form_data[0]
            dependency_media_type = dependency_kwargs_model.expected_form_data[0]
```

File: litestar/dto/factory/backends/msgspec/backend.py [hunk 2]
```diff
@@ -32,6 +32,9 @@ class MsgspecDTOBackend(AbstractDTOBackend[Struct]):
     def parse_raw(self, raw: bytes, media_type: MediaType | str) -> Struct | Collection[Struct]:
         return decode_media_type(raw, media_type, type_=self.annotation)  # type:ignore[no-any-return]
 
+    def populate_data_from_builtins(self, model_type: type[T], data: Any) -> T | Collection[T]:
+        raise NotImplementedError("Msgspec backend does not support marshalling types")
```
Comment (provinzkraut) [inline litestar/dto/factory/backends/msgspec/backend.py:36]: It does support this via `from_builtins`. Maybe a different error message would be appropriate here?

Comment (peterschutt) [reply litestar/dto/factory/backends/msgspec/backend.py:36]: Oh. I assumed from the conversation this morning that there was no coercion of types, but appears that's not the case.

I'm a bit confused with all this now - do I even need the pydantic backend then? I'll experiment with `from_builtins()` tomorrow. Thanks for the heads-up.

Comment (provinzkraut) [reply litestar/dto/factory/backends/msgspec/backend.py:36]: It does, but it's not exactly equivalent to what Pydantic does. I believe the issue was with the passthrough of unsupported types. 

https://jcristharif.com/msgspec/converters.html

Comment (peterschutt) [reply litestar/dto/factory/backends/msgspec/backend.py]: I ended up inmplementing this on the msgspec backed, but still using the pydantic backend for form data for now.

**Comment by peterschutt** [comment]
The missing coverage in this branch needs to be covered in `dto-refactor` branch, not related to this PR.

**Pull request merged**